training/hc-leipzig-7t-mp2rage/plot_nnunet_different_training_logs.py

```diff
@@ -0,0 +1,311 @@
+    Args:
+        log_file_path: Path to the log file.
+    Returns:
+        epoch_and_dice_data: List of dictionaries with keys 'epoch' and 'pseudo_dice'.
+        fold_number: Fold number used for training.
+    """
+    epoch_and_dice_data = []
+    fold_pattern = re.compile(r'Desired fold for training: (\d+)')
+    epoch_pattern = re.compile(r'Epoch (\d+)')
+    dice_pattern = re.compile(r'Pseudo dice \[([^,\]]+(?:, [^,\]]+)*)\]')
+    fold_number = None
+
+    with open(log_file_path, 'r') as file:
+        lines = file.readlines()
+
+    # Extracting the fold number, epoch number and pseudo dice
+    for line in lines:
+        fold_match = fold_pattern.search(line)
+        epoch_match = epoch_pattern.search(line)
+        dice_match = dice_pattern.search(line)
+
+        if fold_match:
+            fold_number = int(fold_match.group(1))
+
+        if epoch_match:
+            epoch = int(epoch_match.group(1))
+            epoch_and_dice_data.append({'epoch': epoch, 'pseudo_dice': None})
+
+        elif dice_match:
+            # Extracting the list using regular expression
+            list_match = re.search(r'\[.*\]', dice_match.group())
+
+            if list_match:
+                extracted_list_str = list_match.group(0)
+                # Replace 'nan' with the actual nan value
+                extracted_list_str = extracted_list_str.replace('nan', 'float("nan")')
+                # Evaluating the string representation of the list
+                extracted_list = eval(extracted_list_str)
+                epoch_and_dice_data[-1]['pseudo_dice'] = extracted_list
+
+    # Check if fold_number is instance of int ("if not fold_number:" does not work for '0')
+    if not isinstance(fold_number, int):
+        fold_number = 'all'
+
+    return epoch_and_dice_data, fold_number
+
+
+def create_figure(df, fname_fig, spinal_level, contrasts):
+    """
+    Create a Seaborn line plot for the validation pseudo dice across epochs and different training logs.
+    :param df: DataFrame containing the epoch number and validation pseudo dice for each class.
+    :param fname_fig: Output figure file path.
+    :param spinal_level: User-defined spinal level for visualization.
+    :param contrasts: Contrast names for the plot title.
+    :return:
+    """
+
+    # Melt the DataFrame for Seaborn plotting
+    df_melted = df.melt(id_vars=['epoch'], var_name='class', value_name='pseudo_dice')
+
+    # Create the Seaborn line plot
+    plt.figure(figsize=(12, 6))
+    sns.set(style="dark", palette="Paired")
+    sns.set_theme()
+
+    # check number of classes - for setting the color palette:
+    num_classes = len(df_melted['class'].unique())
+
+    if num_classes == 2:
+        custom_palette = ["#a6cee3", "#1f78b4"]
+
+    elif num_classes == 3:
+        custom_palette = ["#1f78b4", "#fb9a99", "#e31a1c"]
+
+    elif num_classes == 4:
+        custom_palette = ["#a6cee3", "#1f78b4", "#fb9a99", "#e31a1c"]
```

Comment:
Maybe you can define directly a palette with 4 (or more colours), and seaborn will automatically use only the required number of colours. For example,  in the case of `num_classes == 2`,  seaborn will automatically use only the first two colours. But this needs to be tried.